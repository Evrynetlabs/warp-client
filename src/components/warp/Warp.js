import React, { Component } from 'react'
import Card from 'react-bootstrap/Card'
import classNames from 'classnames'
import WarpChainTitle from 'Components/warp/warp-chain-title/WarpChainTitle'
import WarpContent from 'Components/warp/warp-content/WarpContent'
import { Container, Row, Col } from 'react-bootstrap'
import PropTypes from 'prop-types'
import 'Components/warp/warp.scss'

export default class Warp extends Component {
  constructor(props) {
    super(props)
    this.stylesMain = classNames({
      [this.constructor.name]: true,
    })
    this.stylesContent = `${this.stylesMain}__content`
    this.state = {
      styles: {
        main: this.stylesMain,
        content: this.stylesContent,
      },
    }
  }

  render() {
    return (
      <Container fluid className={this.state.styles.main}>
        <Row>
          <Col>
            <Card className={this.state.styles.content}>
              <WarpChainTitle {...this.props}></WarpChainTitle>
              <WarpContent {...this.props}></WarpContent>
            </Card>
          </Col>
        </Row>
      </Container>
    )
  }
}

Warp.propTypes = {
  txHashes: PropTypes.shape({
    state: PropTypes.shape({
      stellar: PropTypes.string,
      evrynet: PropTypes.string,
    }),
    loading: PropTypes.bool,
    error: PropTypes.object,
  }).isRequired,
  whitelistedAssets: PropTypes.shape({
    state: PropTypes.array,
    loading: PropTypes.bool,
    error: PropTypes.object,
  }).isRequired,
  accountBalance: PropTypes.shape({
    state: PropTypes.string,
    loading: PropTypes.bool,
    error: PropTypes.object,
  }),
  trustlines: PropTypes.shape({
    state: PropTypes.array,
    loading: PropTypes.bool,
    error: PropTypes.object,
  }),
  isToEvrynet: PropTypes.bool.isRequired,
  toEvrynet: PropTypes.func.isRequired,
  toStellar: PropTypes.func.isRequired,
  getWhitelistAssets: PropTypes.func.isRequired,
  toggleTransferSwitch: PropTypes.func.isRequired,
  getAccountBalance: PropTypes.func.isRequired,
  getTrustlines: PropTypes.func.isRequired,
  push: PropTypes.func.isRequired,
  startLoading: PropTypes.func.isRequired,
  stopLoading: PropTypes.func.isRequired,
}
