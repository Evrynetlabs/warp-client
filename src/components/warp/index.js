export { default as WarpComponent } from 'Components/warp/WarpContainer'
export { default as warpReducer } from 'Components/warp/warpReducers'
