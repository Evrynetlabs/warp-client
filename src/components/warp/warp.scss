.Warp {
    height: calc(100% - #{$navbar-height * 2});
    .row {
        height:100%;
        display: flex;
        align-items: center;
        justify-content: center;
    }
    &__content {
        width: 1050px;
        margin: 0 auto;
        &.card {
            color: $light-grey;
            background: none;
            border: none
        }
    }
}