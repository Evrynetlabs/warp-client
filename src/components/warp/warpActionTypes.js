import { createAsyncMutation } from '@/utils/store'

export const ASYNC_COLLECT_HASHES = createAsyncMutation('COLLECT_HASHES')
export const ASYNC_GET_WHITELISTED_ASSETS = createAsyncMutation(
  'GET_WHITELISTED_ASSETS',
)
export const ASYNC_TOGGLE_WARP_SWITCH = createAsyncMutation('IS_TO_EVRY')
export const ASYNC_TOGGLE_GET_ACCOUNT_BALANCE = createAsyncMutation(
  'GET_ACCOUNT_BALANCE',
)

export default {
  ASYNC_COLLECT_HASHES,
  ASYNC_GET_WHITELISTED_ASSETS,
  ASYNC_TOGGLE_WARP_SWITCH,
  ASYNC_TOGGLE_GET_ACCOUNT_BALANCE,
}
