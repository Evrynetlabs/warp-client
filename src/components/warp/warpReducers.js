import actionTypes from 'Components/warp/warpActionTypes'

export const initState = () => {
  return {
    [actionTypes.ASYNC_COLLECT_HASHES.stateKey]: null,
    [actionTypes.ASYNC_COLLECT_HASHES.loadingKey]: true,
    [actionTypes.ASYNC_COLLECT_HASHES.errorKey]: null,
    [actionTypes.ASYNC_GET_WHITELISTED_ASSETS.stateKey]: [],
    [actionTypes.ASYNC_GET_WHITELISTED_ASSETS.loadingKey]: true,
    [actionTypes.ASYNC_GET_WHITELISTED_ASSETS.errorKey]: null,
    [actionTypes.ASYNC_TOGGLE_WARP_SWITCH.stateKey]: true,
    [actionTypes.ASYNC_GET_ACCOUNT_BALANCE.stateKey]: null,
    [actionTypes.ASYNC_GET_ACCOUNT_BALANCE.loadingKey]: true,
    [actionTypes.ASYNC_GET_ACCOUNT_BALANCE.errorKey]: null,
    [actionTypes.ASYNC_GET_TRUSTLINES.stateKey]: [],
    [actionTypes.ASYNC_GET_TRUSTLINES.loadingKey]: true,
    [actionTypes.ASYNC_GET_TRUSTLINES.errorKey]: null,
  }
}

const initialState = initState()

export default function(state = initialState, action) {
  switch (action.type) {
    case actionTypes.ASYNC_COLLECT_HASHES.SUCCESS: {
      const { stellarTxHash, evrynetTxHash } = action.payload
      return {
        ...state,
        [actionTypes.ASYNC_COLLECT_HASHES.stateKey]: {
          stellar: stellarTxHash,
          evrynet: evrynetTxHash,
        },
        [actionTypes.ASYNC_COLLECT_HASHES.loadingKey]: false,
        [actionTypes.ASYNC_COLLECT_HASHES.errorKey]: null,
      }
    }
    case actionTypes.ASYNC_COLLECT_HASHES.PENDING: {
      return {
        ...state,
        [actionTypes.ASYNC_COLLECT_HASHES.stateKey]: null,
        [actionTypes.ASYNC_COLLECT_HASHES.loadingKey]: action.payload,
        [actionTypes.ASYNC_COLLECT_HASHES.errorKey]: null,
      }
    }
    case actionTypes.ASYNC_COLLECT_HASHES.FAILURE: {
      return {
        ...state,
        [actionTypes.ASYNC_COLLECT_HASHES.stateKey]: null,
        [actionTypes.ASYNC_COLLECT_HASHES.loadingKey]: false,
        [actionTypes.ASYNC_COLLECT_HASHES.errorKey]: action.payload,
      }
    }
    case actionTypes.ASYNC_GET_WHITELISTED_ASSETS.SUCCESS: {
      return {
        ...state,
        [actionTypes.ASYNC_GET_WHITELISTED_ASSETS.stateKey]: action.payload,
        [actionTypes.ASYNC_GET_WHITELISTED_ASSETS.loadingKey]: false,
        [actionTypes.ASYNC_GET_WHITELISTED_ASSETS.errorKey]: null,
      }
    }
    case actionTypes.ASYNC_GET_WHITELISTED_ASSETS.PENDING: {
      return {
        ...state,
        [actionTypes.ASYNC_GET_WHITELISTED_ASSETS.stateKey]: [],
        [actionTypes.ASYNC_GET_WHITELISTED_ASSETS.loadingKey]: action.payload,
        [actionTypes.ASYNC_GET_WHITELISTED_ASSETS.errorKey]: null,
      }
    }
    case actionTypes.ASYNC_GET_WHITELISTED_ASSETS.FAILURE: {
      return {
        ...state,
        [actionTypes.ASYNC_GET_WHITELISTED_ASSETS.stateKey]: [],
        [actionTypes.ASYNC_GET_WHITELISTED_ASSETS.loadingKey]: false,
        [actionTypes.ASYNC_GET_WHITELISTED_ASSETS.errorKey]: action.payload,
      }
    }
    case actionTypes.ASYNC_TOGGLE_WARP_SWITCH.SUCCESS: {
      return {
        ...state,
        [actionTypes.ASYNC_TOGGLE_WARP_SWITCH.stateKey]: action.payload,
      }
    }
    case actionTypes.ASYNC_GET_ACCOUNT_BALANCE.PENDING: {
      return {
        ...state,
        [actionTypes.ASYNC_GET_ACCOUNT_BALANCE.stateKey]: null,
        [actionTypes.ASYNC_GET_ACCOUNT_BALANCE.loadingKey]: action.payload,
        [actionTypes.ASYNC_GET_ACCOUNT_BALANCE.errorKey]: null,
      }
    }
    case actionTypes.ASYNC_GET_ACCOUNT_BALANCE.FAILURE: {
      return {
        ...state,
        [actionTypes.ASYNC_GET_ACCOUNT_BALANCE.stateKey]: null,
        [actionTypes.ASYNC_GET_ACCOUNT_BALANCE.loadingKey]: false,
        [actionTypes.ASYNC_GET_ACCOUNT_BALANCE.errorKey]: action.payload,
      }
    }
    case actionTypes.ASYNC_GET_ACCOUNT_BALANCE.SUCCESS: {
      return {
        ...state,
        [actionTypes.ASYNC_GET_ACCOUNT_BALANCE.loadingKey]: false,
        [actionTypes.ASYNC_GET_ACCOUNT_BALANCE.errorKey]: null,
        [actionTypes.ASYNC_GET_ACCOUNT_BALANCE.stateKey]: action.payload,
      }
    }
    case actionTypes.ASYNC_GET_TRUSTLINES.PENDING: {
      return {
        ...state,
        [actionTypes.ASYNC_GET_TRUSTLINES.stateKey]: [],
        [actionTypes.ASYNC_GET_TRUSTLINES.loadingKey]: action.payload,
        [actionTypes.ASYNC_GET_TRUSTLINES.errorKey]: null,
      }
    }
    case actionTypes.ASYNC_GET_TRUSTLINES.FAILURE: {
      return {
        ...state,
        [actionTypes.ASYNC_GET_TRUSTLINES.stateKey]: [],
        [actionTypes.ASYNC_GET_TRUSTLINES.loadingKey]: false,
        [actionTypes.ASYNC_GET_TRUSTLINES.errorKey]: action.payload,
      }
    }
    case actionTypes.ASYNC_GET_TRUSTLINES.SUCCESS: {
      return {
        ...state,
        [actionTypes.ASYNC_GET_TRUSTLINES.loadingKey]: false,
        [actionTypes.ASYNC_GET_TRUSTLINES.errorKey]: null,
        [actionTypes.ASYNC_GET_TRUSTLINES.stateKey]: action.payload,
      }
    }
    default:
      return state
  }
}
