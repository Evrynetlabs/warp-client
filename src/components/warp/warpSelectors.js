import { createStructuredSelector } from 'reselect'
import actionTypes from 'Components/warp/warpActionTypes'

export const isToEvrySelector = (state) =>
  state.warp[actionTypes.ASYNC_TOGGLE_WARP_SWITCH.stateKey]

export const selectCollectedTxHashes = createStructuredSelector({
  state: (state) => {
    return state.warp[actionTypes.ASYNC_COLLECT_HASHES.stateKey]
  },
  loading: (state) => {
    return state.warp[actionTypes.ASYNC_COLLECT_HASHES.loadingKey]
  },
  error: (state) => {
    return state.warp[actionTypes.ASYNC_COLLECT_HASHES.errorKey]
  },
})

export const selectWhitelistedAssets = createStructuredSelector({
  state: (state) => {
    return state.warp[actionTypes.ASYNC_GET_WHITELISTED_ASSETS.stateKey]
  },
  loading: (state) => {
    return state.warp[actionTypes.ASYNC_GET_WHITELISTED_ASSETS.loadingKey]
  },
  error: (state) => {
    return state.warp[actionTypes.ASYNC_GET_WHITELISTED_ASSETS.errorKey]
  },
})

export const selectAccountBalance = createStructuredSelector({
  state: (state) => {
    return state.warp[actionTypes.ASYNC_GET_ACCOUNT_BALANCE.stateKey]
  },
  loading: (state) => {
    return state.warp[actionTypes.ASYNC_GET_ACCOUNT_BALANCE.loadingKey]
  },
  error: (state) => {
    return state.warp[actionTypes.ASYNC_GET_ACCOUNT_BALANCE.errorKey]
  },
})
