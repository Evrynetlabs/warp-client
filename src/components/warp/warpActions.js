import actionTypes from 'Components/warp/warpActionTypes'
import Warp from 'warp-js'
import store from '@/store'

export const collectTxHashesSuccess = ({ stellarTxHash, evrynetTxHash }) => ({
  type: actionTypes.ASYNC_COLLECT_HASHES.SUCCESS,
  payload: { stellarTxHash, evrynetTxHash },
})

export const collectTxHashesPending = (pendingStatus) => ({
  type: actionTypes.ASYNC_COLLECT_HASHES.PENDING,
  payload: pendingStatus,
})

export const collectTxHashesError = (error) => ({
  type: actionTypes.ASYNC_COLLECT_HASHES.FAILURE,
  payload: error,
})

export const toEvry = ({ src, dest, amount, asset }) => {
  const warp = new Warp()
  return async (dispatch) => {
    dispatch(collectTxHashesPending(true))
    try {
      const { stellarTxHash, evrynetTxHash } = await warp.toEvrynet({
        evrynetPriv: dest,
        amount,
        asset,
        stellarPriv: src,
      })
      dispatch(collectTxHashesSuccess({ stellarTxHash, evrynetTxHash }))
    } catch (e) {
      dispatch(collectTxHashesError(e))
    }
  }
}

export const toStellar = ({ src, dest, amount, asset }) => {
  const warp = new Warp()
  return async (dispatch) => {
    dispatch(collectTxHashesPending(true))
    try {
      const { stellarTxHash, evrynetTxHash } = await warp.toStellar({
        evrynetPriv: src,
        amount,
        asset,
        stellarPriv: dest,
      })
      dispatch(collectTxHashesSuccess({ stellarTxHash, evrynetTxHash }))
    } catch (e) {
      dispatch(collectTxHashesError(e))
    }
  }
}

export const getWhitelistAssetsSuccess = (whitelistedAssets) => ({
  type: actionTypes.ASYNC_GET_WHITELISTED_ASSETS.SUCCESS,
  payload: whitelistedAssets,
})

export const getWhitelistAssetsPending = (pendingStatus) => ({
  type: actionTypes.ASYNC_GET_WHITELISTED_ASSETS.PENDING,
  payload: pendingStatus,
})

export const getWhitelistAssetsError = (error) => ({
  type: actionTypes.ASYNC_GET_WHITELISTED_ASSETS.FAILURE,
  payload: error,
})

export const getWhitelistAssets = () => {
  const warp = new Warp()
  return async (dispatch) => {
    dispatch(getWhitelistAssetsPending(true))
    try {
      const whitelistedAssets = await warp.client.evry.getWhitelistAssets()
      dispatch(getWhitelistAssetsSuccess(whitelistedAssets.assets))
    } catch (e) {
      dispatch(getWhitelistAssetsError(e))
    }
  }
}

export const toggleTransferSwitch = () => ({
  type: actionTypes.ASYNC_TOGGLE_WARP_SWITCH.SUCCESS,
  payload: !store.getState().warp[
    actionTypes.ASYNC_TOGGLE_WARP_SWITCH.stateKey
  ],
})
