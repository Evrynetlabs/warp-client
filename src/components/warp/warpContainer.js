import { connect } from 'react-redux'
import { compose } from 'redux'
import Warp from 'Components/warp/Warp'
import {
  isToEvrySelector,
  selectCollectedTxHashes,
  selectWhitelistedAssets,
  selectAccountBalance,
} from 'Components/warp/warpSelectors'
import {
  toEvrynet,
  toStellar,
  getWhitelistAssets,
  toggleTransferSwitch,
  getAccountBalance,
} from 'Components/warp/warpActions'

const mapStateToProps = (state) => ({
  isToEvry: isToEvrySelector(state),
  txHashes: selectCollectedTxHashes(state),
  whitelistedAssets: selectWhitelistedAssets(state),
  accountBalance: selectAccountBalance(state),
})

const mapDispatchToProps = (dispatch) => {
  return {
    toEvrynet: (payload) => dispatch(toEvrynet(payload)),
    toStellar: (payload) => dispatch(toStellar(payload)),
    getWhitelistAssets: () => dispatch(getWhitelistAssets()),
    toggleTransferSwitch: () => dispatch(toggleTransferSwitch()),
    getAccountBalance: (payload) => dispatch(getAccountBalance(payload)),
  }
}

const withConnect = connect(
  mapStateToProps,
  mapDispatchToProps,
)

export default compose(withConnect)(Warp)
