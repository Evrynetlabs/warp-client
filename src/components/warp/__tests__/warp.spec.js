import React from 'react'
import Enzyme, { shallow } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import WarpContainer from 'Components/warp/warpContainer'
import configureStore from 'redux-mock-store'
import thunk from 'redux-thunk'
import { initState } from 'Components/warp/warpReducers'

Enzyme.configure({ adapter: new Adapter() })

describe('Warp', () => {
  const mockStore = configureStore([thunk])
  let store, container

  beforeEach(() => {
    store = mockStore({
      warp: initState(),
    })
    container = shallow(<WarpContainer store={store} />)
  })
  it('should render correspondingly', () => {
    expect(container).toMatchSnapshot()
  })
})
