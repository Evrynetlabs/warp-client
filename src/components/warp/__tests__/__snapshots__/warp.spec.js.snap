// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Warp should render correspondingly 1`] = `
<ContextProvider
  value={null}
>
  <Warp
    getWhitelistAssets={[Function]}
    isToEvry={true}
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
    toEvry={[Function]}
    toStellar={[Function]}
    toggleTransferSwitch={[Function]}
    txHashes={
      Object {
        "error": null,
        "loading": true,
        "state": null,
      }
    }
    whitelistedAssets={
      Object {
        "error": null,
        "loading": true,
        "state": Array [],
      }
    }
  />
</ContextProvider>
`;
